255 calendar weekly UI

Harishmakka opened the pull request:
feat: Implement WeeklyView with CalendarDayCard and add comprehensive tests

•⁠  ⁠Integrated CalendarDayCard into WeeklyView for consistent day rendering
•⁠  ⁠Memoized evaluations per day based on selected start date
•⁠  ⁠Displayed fallback message for days with no evaluations
•⁠  ⁠Added test cases to verify:
  - Correct rendering of 7 day columns
  - Display of evaluation titles on correct days
  - Presence of fallback messages using getAllByText
•⁠  ⁠Aligned test assertions with feedback for accuracy and clarity

Harishmakka committed:
feat: Implement WeeklyView with global evaluation, CalendarDayCard and add comprehensive tests

- Integrated CalendarDayCard into WeeklyView for consistent day rendering
- Memoized evaluations per day based on selected start date
- Displayed fallback message for days with no evaluations
- Added test cases to verify:
  - Correct rendering of 7 day columns
  - Display of evaluation titles on correct days
  - Presence of fallback messages using getAllByText
- Aligned test assertions with feedback for accuracy and clarity

Harishmakka committed:
tests

achow-conestogac commented on src/EvaluationCalendar/WeeklyView/WeeklyView.tsx:
```diff
@@ -0,0 +1,38 @@
+
+interface WeeklyViewProps {
+  evaluations: Evaluation[];
+  startDate: Date;
```
what does `startDate` do?

does the start date need to match one of the evaluations? why do you need a start date? shouldn't the weekly view simply display the `evaluations` chronologically?

Harishmakka committed:
feat: add CalendarPdfExportButtons component with tests

- Adds dropdown-based Export to PDF button
- Handles export options for Daily, Weekly, Monthly, and Course
- Includes unit tests verifying UI behavior and handler calls

Harishmakka committed:
Merge branch 'main' into 255-CalendarWeeklyUI

Harishmakka committed:
refactor: remove startDate prop from WeeklyView and update tests

achow-conestogac commented:
okay. so where is the usage of this component?

Harishmakka committed:
integrate WeeklyView into CalendarView with viewMode support

Harishmakka commented:
Sorry Professor,the changes were committed locally but I hadn’t pushed them yet. I’ve just pushed the 255-CalendarWeeklyUI branch now. The WeeklyView is properly integrated in CalendarView.tsx and should be visible on GitHub. Please let me know if it shows up on your end now.

Harishmakka committed:
Merge branch 'main' into 255-CalendarWeeklyUI

Harishmakka committed:
merge main with working branch and resolve the conflicts and update case cases

Harishmakka committed:
refactor: update tests

Harishmakka committed:
remove comments

Harishmakka committed:
fix: resolve module loading issue by reinstalling dependencies

Harishmakka commented:
fix: resolve module loading issue by reinstalling dependencies
npm run dev is working now

Harishmakka committed:
fix: improve CalendarView fallback rendering, update navigation labels, and enhance tests

Harishmakka committed:
feat: Integrated WeeklyView into CalendarView safely

- Preserved existing CalendarView logic and structure
- Ensured no changes were made to the original navigation logic

Harishmakka commented on src/Evaluation/EvaluationCalendar/CalendarView/CalendarView.tsx:
```diff
@@ -43,27 +45,27 @@ const CalendarView: React.FC<CalendarViewProps> = ({ evaluations }) => {
         onPrev={() =>
-          view === "weekly"
-            ? navigateWeek("prev")
-            : navigateMonth("prev")
+          view === "weekly" ? navigateWeek("prev") : navigateMonth("prev")
         }
```
Thanks for the feedback, Professor, with the recent commit I Preserved existing CalendarView logic and structure
- Ensured no changes were made to the original navigation logic

Harishmakka committed:
Removed comments

Harishmakka committed:
refactor(WeeklyView): optimize performance and cleanup constants

- Moved DATE_FORMAT constant outside of component to avoid redundant re-creation
- Memoized startOfWeek and daysOfWeek using useMemo for efficient rendering
- Applied useMemo to dailyEvaluations filter to optimize evaluation lookup

Harishmakka commented on src/Evaluation/EvaluationCalendar/WeeklyView/WeeklyView.test.tsx:
```diff
@@ -0,0 +1,81 @@
+import { Evaluation } from "../../EvaluationService";
+
+describe("WeeklyView", () => {
+  // MATCHING what the component renders
```
removed the comments.

Harishmakka commented on src/Evaluation/EvaluationCalendar/WeeklyView/WeeklyView.tsx:
```diff
@@ -0,0 +1,70 @@
+    return (
+        <div className="grid grid-cols-7 gap-4">
+            {daysOfWeek.map((day) => {
+                const dateStr = format(day, "EEE MMM dd yyyy");
```
Moved the DATE_FORMAT constant outside the component scope, just below the interface, to avoid redundant re-creation during re-renders.

Harishmakka commented on src/Evaluation/EvaluationCalendar/WeeklyView/WeeklyView.tsx:
```diff
@@ -0,0 +1,70 @@
+                const dateStr = format(day, "EEE MMM dd yyyy");
+                const dailyEvaluations = evaluations.filter(
+                    (ev) => format(ev.dueDate, "EEE MMM dd yyyy") === dateStr
+                );
```
Applied useMemo() to memoize the dailyEvaluations filter operation so that it recalculates only when dependencies change.

Harishmakka committed:
Merge branch 'main' into 255-CalendarWeeklyUI

achow-conestogac commented on src/Evaluation/EvaluationCalendar/WeeklyView/WeeklyView.tsx:
```diff
@@ -0,0 +1,79 @@
+
+interface WeeklyViewProps {
+    evaluations: Evaluation[];
+    currentWeekStart: Date;
```
as we said in class, why not simply use the `evaluation` data in order to deal with this?

at most you need people to day `"monday"` or `"sunday"`.

Harishmakka replied:
Thank you for the clarification! I’ve removed the currentWeekStart prop and now derive the week's start directly from the earliest evaluation in the list. The view defaults to Monday-based weeks, as suggested. Please let me know if further refinements are needed!

Harishmakka committed:
Merge branch 'main' into 255-CalendarWeeklyUI

Harishmakka committed:
resolved conflicts and ensured all tests are passed

Harishmakka committed:
efactor WeeklyView to derive week start from evaluation data

- Removed `currentWeekStart` prop as per professor's feedback
- Week now starts on Monday based on earliest evaluation date
- Simplified props and improved component logic
- Updated tests to reflect the new weekly grouping logic

achow-conestogac commented on src/Evaluation/EvaluationCalendar/CalendarView/CalendarView.test.tsx:
```diff
@@ -34,62 +34,79 @@ describe("CalendarView", () => {
   ];
 
-  it("renders weekly view with evaluation cards by default", () => {
+  beforeAll(() => {
+    jest.useFakeTimers().setSystemTime(new Date("2025-06-24T12:00:00-04:00"));
+  });
+
+  afterAll(() => {
+    jest.useRealTimers();
+  });
```
as i have already said, please do not use mocking to deal with this.

The Date library has a ton of ways for you to set Dates.  you need to construct your Date dynamically.

Harishmakka replied:
Thank you for the clarification. I've removed all usages of jest.useFakeTimers() and now rely entirely on constructing Date instances directly in tests. All test cases now use explicit dueDate values to ensure consistency without mocking time. Let me know if this is aligned with your expectations or if any further changes are needed.

Harishmakka replied:
I’ve revised the tests to construct all test Date objects explicitly based on dynamic logic without using jest.useFakeTimers.

I’ve also improved the test data to better reflect realistic weekly distributions, ensuring the test coverage aligns with actual use cases.

achow-conestogac reviewed (commented):
this implementation is very rough. 

@Harishmakka you need to do a better job with the test data. 

You shouldn't need to force the system to be a certain date in order to get your tests to work.

Harishmakka committed:
refactor(tests): remove jest.useFakeTimers and construct Dates explicitly

achow-conestogac commented on src/Evaluation/EvaluationCalendar/WeeklyView/WeeklyView.tsx:
```diff
@@ -0,0 +1,83 @@
+const WeeklyView: React.FC<WeeklyViewProps> = ({ evaluations }) => {
+  const startOfWeek = useMemo(() => {
+    if (evaluations.length === 0) {
+      return getStartOfWeek(new Date()); // fallback to current week
```
this is an example of a good comment. but please be more elaborate.

```suggestion
      return getStartOfWeek(new Date()); // fallback to current week if no evaluation is found.
```

this might seem obvious right now, but as the code grows, this forms a direct connection between the if check and what you are returning.

Harishmakka committed:
Update src/Evaluation/EvaluationCalendar/WeeklyView/WeeklyView.tsx

Co-authored-by: Andy Chow <168104245+achow-conestogac@users.noreply.github.com>

Harishmakka commented:
Hi Professor, Thank you for the Suggestion. Updated according to it.
